oss-internship-2020/libtiff/example/main_sandboxed.cc

```diff
@@ -0,0 +1,325 @@
+      TIFFGetR(rgba) <= static_cast<unsigned>(limits.max_red) &&
+      TIFFGetG(rgba) >= static_cast<unsigned>(limits.min_green) &&
+      TIFFGetG(rgba) <= static_cast<unsigned>(limits.max_green) &&
+      TIFFGetB(rgba) >= static_cast<unsigned>(limits.min_blue) &&
+      TIFFGetB(rgba) <= static_cast<unsigned>(limits.max_blue) &&
+      TIFFGetA(rgba) >= static_cast<unsigned>(limits.min_alpha) &&
+      TIFFGetA(rgba) <= static_cast<unsigned>(limits.max_alpha)) {
+    return absl::OkStatus();
+  }
+
+  return absl::InternalError(absl::StrFormat(
+      kRgbaPixelErrorFormatStr, pixel, TIFFGetR(rgba), limits.min_red,
+      limits.max_red, TIFFGetG(rgba), limits.min_green, limits.max_green,
+      TIFFGetB(rgba), limits.min_blue, limits.max_blue, TIFFGetA(rgba),
+      limits.min_alpha, limits.max_alpha));
+}
+
+}  // namespace
+
+std::string GetFilePath(const std::string& dir, const std::string& filename) {
+  return sandbox2::file::JoinPath(dir, "test", "images", filename);
+}
+
+std::string GetFilePath(const std::string filename) {
+  std::string cwd = sandbox2::file_util::fileops::GetCWD();
+  auto find = cwd.rfind("build");
+
+  std::string project_path;
+  if (find == std::string::npos) {
+    LOG(ERROR)
+        << "Something went wrong: CWD don't contain build dir. "
+        << "Please run tests from build dir or send project dir as a "
+        << "parameter: ./sandboxed /absolute/path/to/project/dir .\n"
+        << "Falling back to using current working directory as root dir.\n";
+    project_path = cwd;
+  } else {
+    project_path = cwd.substr(0, find);
+  }
+
+  return sandbox2::file::JoinPath(project_path, "test", "images", filename);
+}
+
+absl::Status LibTIFFMain(const std::string& srcfile) {
+  // to use dir and file inside sapi-libtiff, use
+  // sandbox(file) – file only -- or
+  // sandbox(file, dir) -- file and dir -- or
+  // sandbox(nullopt, dir) -- dir only.
+  // file and directory must exist.
+  // all paths must be absolute.
+
+  TiffSapiSandbox sandbox(srcfile);
+
+  bool pixel_status_ok = true;
+  bool cluster_status_ok = true;
+  // initialize sapi vars after constructing TiffSapiSandbox
+  sapi::v::UShort h;
+  sapi::v::UShort v;
+  absl::StatusOr<TIFF*> status_or_tif;
+  absl::StatusOr<int> status_or_int;
+  absl::StatusOr<tmsize_t> status_or_long;
+  absl::Status status;
+
+  status = sandbox.Init();
+
+  SAPI_RETURN_IF_ERROR(sandbox.Init());
+
+  TiffApi api(&sandbox);
+  sapi::v::ConstCStr srcfile_var(srcfile.c_str());
+  sapi::v::ConstCStr r_var("r");
+
+  SAPI_ASSIGN_OR_RETURN(
+      status_or_tif, api.TIFFOpen(srcfile_var.PtrBefore(), r_var.PtrBefore()));
+
+  sapi::v::RemotePtr tif(status_or_tif.value());
+  if (!tif.GetValue()) {
+    return absl::InternalError(absl::StrCat("Could not open ", srcfile));
+  }
+
+  SAPI_ASSIGN_OR_RETURN(auto return_value,
+                        api.TIFFGetField2(&tif, TIFFTAG_YCBCRSUBSAMPLING,
+                                          h.PtrBoth(), v.PtrBoth()));
```

Comment:
`PtrAfter` should be enough if `TIFFGetField` is just a getter.
It will still allocate storage for the vars prior to calling the function.